File: sdk/basin-dotnet/tests/Basin.Tests/StorageTests.cs
using System.Net;
using System.Text;
using System.Text.Json;
using Xunit;

namespace Basin.Tests;

/// <summary>
/// StorageClient and StorageBucketClient tests.
/// </summary>
public class StorageTests
{
    private const string BaseUrl = "http://localhost";

    private static BasinClient BuildClient(MockHandler handler, string? projectId = "proj-001") =>
        new BasinClient.Builder()
            .WithUrl(BaseUrl)
            .WithApiKey("test-key")
            .WithProjectId(projectId)
            .WithHttpClient(new HttpClient(handler))
            .Build();

    private static object BucketPayload(string name = "avatars") => new
    {
        id = "bucket-001",
        name,
        @public = false,
        file_size_limit = (long?)null,
        allowed_mime_types = new string[] { },
        created_at = "2024-01-01T00:00:00Z",
        updated_at = "2024-01-01T00:00:00Z",
    };

    private static object ObjectPayload(string path = "alice/photo.png") => new
    {
        id = "obj-001",
        bucket_id = "bucket-001",
        path,
        size = 1024L,
        mime_type = "image/png",
        metadata = new { },
        owner = "user-001",
        etag = "\"abc123\"",
        created_at = "2024-01-01T00:00:00Z",
        updated_at = "2024-01-01T00:00:00Z",
    };

    // ------------------------------------------------------------------
    // Bucket management
    // ------------------------------------------------------------------

    [Fact]
    public async Task CreateBucketAsync_ReturnsBucket()
    {
        var handler = new MockHandler()
            .RespondJson(Match.Post("/storage/v1/bucket"), BucketPayload(), HttpStatusCode.Created);

        using var client = BuildClient(handler);
        var bucket = await client.Storage.CreateBucketAsync("avatars", @public: true);

        Assert.Equal("avatars", bucket.Name);
        Assert.Equal("bucket-001", bucket.Id);
    }

    [Fact]
    public async Task GetBucketAsync_ReturnsBucket()
    {
        var handler = new MockHandler()
            .RespondJson(Match.Get("/storage/v1/bucket/avatars"), BucketPayload());

        using var client = BuildClient(handler);
        var bucket = await client.Storage.GetBucketAsync("avatars");
        Assert.Equal("avatars", bucket.Name);
    }

    [Fact]
    public async Task DeleteBucketAsync_CallsDeleteRoute()
    {
        var handler = new MockHandler()
            .RespondJson(Match.Delete("/storage/v1/bucket/avatars"), new { ok = true });

        using var client = BuildClient(handler);
        await client.Storage.DeleteBucketAsync("avatars");

        Assert.Equal(HttpMethod.Delete, handler.Requests.Single().Method);
    }

    // ------------------------------------------------------------------
    // Object operations
    // ------------------------------------------------------------------

    [Fact]
    public async Task UploadAsync_PostsBytesToObjectRoute()
    {
        var handler = new MockHandler()
            .RespondJson(Match.Post("/storage/v1/object/avatars"), ObjectPayload());

        using var client = BuildClient(handler);
        var obj = await client.Storage.FromBucket("avatars")
                              .UploadAsync("alice/photo.png", new byte[] { 1, 2, 3 }, "image/png");

        Assert.Equal("alice/photo.png", obj.Path);
        Assert.Equal(HttpMethod.Post, handler.Requests.Single().Method);
    }

    [Fact]
    public async Task DownloadAsync_ReturnsBytesAndHeaders()
    {
        var handler = new MockHandler();
        handler.Respond(
            Match.Get("/storage/v1/object/avatars"),
            HttpStatusCode.OK,
            body: null,   // will be set below via content
            contentType: "image/png");

        // Rebuild mock with actual content
        var handler2 = new MockHandler();
        var content = new byte[] { 0x89, 0x50, 0x4E, 0x47 }; // PNG magic bytes
        handler2.Respond(
            Match.Get("/storage/v1/object/avatars"),
            HttpStatusCode.OK,
            body: Convert.ToBase64String(content), // just need non-empty
            contentType: "image/png");

        using var client = BuildClient(handler2);
        var result = await client.Storage.FromBucket("avatars").DownloadAsync("alice/photo.png");

        Assert.NotNull(result.Data);
        Assert.Equal("image/png", result.ContentType);
    }

    [Fact]
    public async Task RemoveAsync_DeletesSingleObject()
    {
        var handler = new MockHandler()
            .RespondJson(Match.Delete("/storage/v1/object/avatars/alice/photo.png"), new { ok = true });

        using var client = BuildClient(handler);
        await client.Storage.FromBucket("avatars").RemoveAsync("alice/photo.png");

        Assert.Equal(HttpMethod.Delete, handler.Requests.Single().Method);
    }

    [Fact]
    public async Task ListAsync_PostsToListRoute()
    {
        var handler = new MockHandler()
            .RespondJson(Match.Post("/storage/v1/object/list/avatars"),
                new[] { ObjectPayload() });

        using var client = BuildClient(handler);
        var objects = await client.Storage.FromBucket("avatars").ListAsync(prefix: "alice/");

        Assert.Single(objects);
        Assert.Equal("alice/photo.png", objects[0].Path);
    }

    [Fact]
    public async Task RemoveByPrefixesAsync_DeletesWithBody()
    {
        var handler = new MockHandler()
            .RespondJson(Match.Delete("/storage/v1/object/avatars"), new { ok = true });

        using var client = BuildClient(handler);
        await client.Storage.FromBucket("avatars")
                    .RemoveByPrefixesAsync(new[] { "alice/", "bob/" });

        Assert.Equal(HttpMethod.Delete, handler.Requests.Single().Method);
    }

    // ------------------------------------------------------------------
    // Public URL + signed URL
    // ------------------------------------------------------------------

    [Fact]
    public void GetPublicUrl_BuildsCorrectUrl()
    {
        var handler = new MockHandler();
        using var client = BuildClient(handler, "proj-001");

        var url = client.Storage.FromBucket("avatars").GetPublicUrl("alice/photo.png");
        Assert.Equal("http://localhost/storage/v1/object/public/proj-001/avatars/alice/photo.png", url);
    }

    [Fact]
    public void GetPublicUrl_ThrowsWhenNoProjectId()
    {
        var handler = new MockHandler();
        using var client = BuildClient(handler, projectId: null);

        var ex = Assert.Throws<BasinApiException>(() =>
            client.Storage.FromBucket("avatars").GetPublicUrl("file.txt"));

        Assert.Equal("E_INVALID_REQUEST", ex.Code);
    }

    [Fact]
    public async Task CreateSignedUrlAsync_ReturnsSignedUrl()
    {
        // Wire response uses camelCase: signedUrl / expiresAt (storage_sign.rs)
        var handler = new MockHandler()
            .RespondJson(
                Match.Post("/storage/v1/object/sign/upload/avatars"),
                new
                {
                    signedUrl = "/storage/v1/object/sign/proj-001/avatars/alice/photo.png?token=xyz",
                    expiresAt = "2099-01-01T00:00:00Z",
                });

        using var client = BuildClient(handler);
        var signed = await client.Storage.FromBucket("avatars")
                                 .CreateSignedUrlAsync("alice/photo.png", 3600);

        Assert.Contains("token=xyz", signed.SignedUrlPath);
        Assert.Equal("2099-01-01T00:00:00Z", signed.ExpiresAt);
        Assert.StartsWith("http://localhost", signed.AbsoluteUrl);
    }
}

File: sdk/basin-swift/Tests/BasinTests/StorageTests.swift
import XCTest
@testable import Basin

final class StorageTests: XCTestCase {

    var client: BasinClient!

    override func setUp() {
        super.setUp()
        MockURLProtocol.reset()
        client = BasinClient(url: "http://localhost:8080", key: "key", projectID: "proj_01J", session: mockSession())
    }

    private let bucketJSON: [String: Any] = [
        "id": "bkt_001", "name": "avatars", "public": true,
        "file_size_limit": NSNull(), "allowed_mime_types": [],
        "created_at": "2024-01-01T00:00:00Z", "updated_at": "2024-01-01T00:00:00Z"
    ]

    private let objectJSON: [String: Any] = [
        "id": "obj_001", "bucket_id": "bkt_001", "path": "users/alice.png",
        "size": 1024, "mime_type": "image/png", "owner": NSNull(), "etag": "abc123",
        "created_at": "2024-01-01T00:00:00Z", "updated_at": "2024-01-01T00:00:00Z",
        "metadata": NSNull()
    ]

    // MARK: - Bucket operations

    func testCreateBucket() async throws {
        MockURLProtocol.push { req in
            XCTAssertEqual(req.httpMethod, "POST")
            XCTAssertEqual(req.url?.path, "/storage/v1/bucket")
            return mockResponse(url: req.url!, status: 201, json: self.bucketJSON)
        }
        let bucket = try await client.storage.createBucket("avatars", public: true)
        XCTAssertEqual(bucket.name, "avatars")
        XCTAssertTrue(bucket.public)
    }

    func testGetBucket() async throws {
        MockURLProtocol.push { req in
            XCTAssertEqual(req.httpMethod, "GET")
            XCTAssert(req.url?.path.contains("/storage/v1/bucket/") == true)
            return mockResponse(url: req.url!, status: 200, json: self.bucketJSON)
        }
        let bucket = try await client.storage.getBucket("avatars")
        XCTAssertEqual(bucket.id, "bkt_001")
    }

    func testDeleteBucket() async throws {
        MockURLProtocol.push { req in
            XCTAssertEqual(req.httpMethod, "DELETE")
            return mockResponse(url: req.url!, status: 200, json: ["ok": true])
        }
        _ = try await client.storage.deleteBucket("avatars")
    }

    // MARK: - Object operations

    func testUpload() async throws {
        MockURLProtocol.push { req in
            XCTAssertEqual(req.httpMethod, "POST")
            XCTAssert(req.url?.path.contains("/storage/v1/object/avatars/") == true)
            return mockResponse(url: req.url!, status: 201, json: self.objectJSON)
        }
        let bucket = await client.storage.fromBucket("avatars")
        let obj = try await bucket.upload("users/alice.png",
                                          data: Data([0x89, 0x50, 0x4E, 0x47]),
                                          contentType: "image/png")
        XCTAssertEqual(obj.path, "users/alice.png")
        XCTAssertEqual(obj.size, 1024)
    }

    func testDownload() async throws {
        let imageData = Data([0x89, 0x50, 0x4E, 0x47])
        MockURLProtocol.push { req in
            XCTAssertEqual(req.httpMethod, "GET")
            let response = HTTPURLResponse(
                url: req.url!, statusCode: 200, httpVersion: nil,
                headerFields: ["Content-Type": "image/png", "ETag": "abc123"]
            )!
            return (imageData, response)
        }
        let bucket = await client.storage.fromBucket("avatars")
        let result = try await bucket.download("users/alice.png")
        XCTAssertEqual(result.data, imageData)
        XCTAssertEqual(result.contentType, "image/png")
        XCTAssertEqual(result.etag, "abc123")
    }

    func testRemoveObject() async throws {
        MockURLProtocol.push { req in
            XCTAssertEqual(req.httpMethod, "DELETE")
            return mockResponse(url: req.url!, status: 200, json: ["ok": true])
        }
        let bucket = await client.storage.fromBucket("avatars")
        _ = try await bucket.remove("users/alice.png")
    }

    func testListObjects() async throws {
        MockURLProtocol.push { req in
            XCTAssertEqual(req.httpMethod, "POST")
            XCTAssert(req.url?.path.contains("/storage/v1/object/list/avatars") == true)
            return mockResponse(url: req.url!, status: 200, json: [self.objectJSON])
        }
        let bucket = await client.storage.fromBucket("avatars")
        let objects = try await bucket.list(prefix: "users/")
        XCTAssertEqual(objects.count, 1)
        XCTAssertEqual(objects[0].etag, "abc123")
    }

    func testRemoveByPrefixes() async throws {
        MockURLProtocol.push { req in
            XCTAssertEqual(req.httpMethod, "DELETE")
            XCTAssert(req.url?.path.hasSuffix("/storage/v1/object/avatars") == true)
            return mockResponse(url: req.url!, status: 200, json: ["ok": true])
        }
        let bucket = await client.storage.fromBucket("avatars")
        _ = try await bucket.removeByPrefixes(["users/alice.png", "users/bob.png"])
    }

    // MARK: - URLs

    func testGetPublicURL() async throws {
        let bucket = await client.storage.fromBucket("avatars")
        let url = try await bucket.getPublicURL("users/alice.png")
        XCTAssert(url.contains("/storage/v1/object/public/proj_01J/avatars/users/alice.png"), "got: \(url)")
    }

    func testGetPublicURL_NoProjectID_Throws() async {
        let clientNoProject = BasinClient(url: "http://localhost:8080", key: "key", session: mockSession())
        let bucket = await clientNoProject.storage.fromBucket("avatars")
        do {
            _ = try await bucket.getPublicURL("test.png")
            XCTFail("Expected BasinApiError")
        } catch let e as BasinApiError {
            XCTAssertEqual(e.rawCode, "E_INVALID_REQUEST")
        } catch {
            XCTFail("Unexpected error: \(error)")
        }
    }

    func testErrorEnvelopeSurfacesSQLState() throws {
        let body = #"{"code":"E_INVALID_REQUEST","message":"duplicate key value violates unique constraint","sqlstate":"23505"}"#
        let err = BasinApiError.fromResponseBody(Data(body.utf8), status: 400)
        XCTAssertEqual(err.rawCode, "E_INVALID_REQUEST")
        XCTAssertEqual(err.sqlState, "23505")
        XCTAssert(err.description.contains("23505"))
    }

    func testErrorEnvelopeSQLStateNilWhenAbsent() throws {
        let body = #"{"code":"E_NOT_FOUND","message":"gone"}"#
        let err = BasinApiError.fromResponseBody(Data(body.utf8), status: 404)
        XCTAssertNil(err.sqlState)
    }

    func testCreateSignedURL() async throws {
        MockURLProtocol.push { req in
            XCTAssertEqual(req.httpMethod, "POST")
            XCTAssert(req.url?.path.contains("/storage/v1/object/sign/upload/avatars/") == true)
            // Server returns camelCase per storage_sign.rs.
            return mockResponse(url: req.url!, status: 200, json: [
                "signedUrl": "/storage/v1/object/sign/proj_01J/avatars/users/alice.png?token=sig_tok",
                "expiresAt": "2099-01-01T00:00:00Z"
            ])
        }
        let bucket = await client.storage.fromBucket("avatars")
        let signed = try await bucket.createSignedURL("users/alice.png", expiresIn: 3600)
        XCTAssert(signed.absoluteUrl.hasPrefix("http://localhost:8080"), "got: \(signed.absoluteUrl)")
        XCTAssert(signed.absoluteUrl.contains("sig_tok"), "got: \(signed.absoluteUrl)")
        XCTAssertFalse(signed.expiresAt.isEmpty)
    }
}
